Annotate required APIs with DynamicallyAccessedMemberTypes.Interfaces

vitek-karas opened the pull request:
After the basic annotation of Type.GetInterface and Type.GetInterfaces this propagates this annotation to all places which require it. It also resolves warnings in all the places where this shows up.

For the most part this is just blindly propagating the annotations.

Most of the suppressions rely on the fact that is `IFoo` is referenced (and thus the `Type` instance of it exists at runtime), trimmer guarantees that it will keep it on all types which implement that interface. And also on additional fact that if `IFoo<>` is preserved, then all instantiations of it are also preserved. This is true for linker, but not necessarily true for AOT - that is we can do better if we don't have to make that assumption. For now I rely on that assumption.

vitek-karas committed:
Annotate required APIs with DynamicallyAccessedMemberTypes.Interfaces

After the basic annotation of Type.GetInterface and Type.GetInterfaces this propagates this annotation to all places which require it. It also resolves warnings in all the places where this shows up.

vitek-karas committed:
Fix build

MichalStrehovsky commented on src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:
```diff
@@ -803,6 +803,10 @@ private RuntimeConstructorInfo[] PopulateConstructors(Filter filter)
                     return list.ToArray();
                 }
 
+                [UnconditionalSuppressMessage ("ReflectionAnalysis", "IL2075:UnrecognizedReflectionPattern",
```
I've submitted https://github.com/dotnet/runtime/pull/52535 to see whether we can just delete the problematic code instead.

I think this would only do something if we allowed instance fields on interfaces, but that smells a lot like multiple-inheritance problems, so I don't think we're ever going to allow that.

vitek-karas replied:
Thanks a lot for testing this.

MichalStrehovsky commented on src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:
```diff
@@ -1003,6 +1011,10 @@ private void AddSpecialInterface(ref ListBuilder<RuntimeType> list, Filter filte
                     }
                 }
 
+                [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2065:UnrecognizedReflectionPattern",
```
Can we put `[DynamicallyAccessed(Interfaces)]` on the method?

vitek-karas replied:
It's very tricky. This all rolls up to `RuntimeTypeCache.m_runtimeType` which would have to be annotated, but that is called for example from `Type.GetProperties` which is annotated with `Properties`. The suppression would have to be somewhere. This all boils down to the fact that the `RuntimeTypeCache` caches everything in one class - we fill it piecemeal (if fields are needed we fill fields, later on we may add methods and so on), but statically it can contain all of the member types at once, and so it would need to be annotated as such.

The reason why this is the first time we hit this is that we didn't annotate the runtime calls in reflection as "dangerous - for example RuntimeTypeHandle.GetNextIntroducedMethod should be "dangerous" as in requiring all methods on the type. But annotating all of this is tricky and will eventually all clash inside the `RuntimeTypeCache` anyway.

MichalStrehovsky commented on src/libraries/System.Private.Xml/src/System/Xml/Serialization/CodeGenerator.cs:
```diff
@@ -53,6 +53,10 @@ internal static bool IsNullableGenericType(Type type)
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "Debug only code, we don't ship debug binaries.")]
+#endif
         internal static void AssertHasInterface(Type type, Type iType)
```
Nit: Can we put `[Conditional("DEBUG")]` on the method itself to eliminate all the ifdeffing?

vitek-karas committed:
Merge remote-tracking branch 'origin/main' into GetInterfacesAnnotation

vitek-karas committed:
Merge with main and fix warnings in OOB packages

eerhardt commented on src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:
```diff
@@ -159,6 +159,16 @@ private static IEnumerable<Type> FindConfigurationServices(Type type)
+                    "The trimmer will keep the interface and thus all of its implementations in that case. " +
+                    "The call to GetInterfaces may return less results in trimmed apps, but it will " +
+                    "include the interfaces this method looks for if they should be there.")]
+            static Type[] GetInterfacesOnType (Type t)
```
```suggestion
            static Type[] GetInterfacesOnType(Type t)
```

MichalStrehovsky replied:
We need to make this part of the process after working in the linker repo:

<img src="https://blog.gosafe.com/hs-fs/hubfs/Campaign_1_Facility_Safety/C1_Blog%20Images/Eyewash%20Gif%203.gif?width=1024&name=Eyewash%20Gif%203.gif">

eerhardt commented on src/libraries/System.Private.CoreLib/src/System/Reflection/SignatureType.cs:
```diff
@@ -173,7 +176,10 @@ public sealed override Type MakeArrayType(int rank)
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.Interfaces)]
         public sealed override Type[] FindInterfaces(TypeFilter filter, object? filterCriteria) => throw new NotSupportedException(SR.NotSupported_SignatureType);
+
         public sealed override InterfaceMapping GetInterfaceMap(Type interfaceType) => throw new NotSupportedException(SR.NotSupported_SignatureType);
```
What about `GetInterfaceMap`? Does this need any annotations?

eerhardt replied:
> Maybe we need to annotate it with PublicMethods

Does `PublicMethods` include explicitly implemented methods?

MichalStrehovsky replied:
Might have to be all the methods on the interface: https://github.com/dotnet/runtime/issues/46912#issuecomment-837179543

eerhardt commented on src/libraries/System.Reflection.TypeExtensions/ref/System.Reflection.TypeExtensions.cs:
```diff
@@ -61,7 +61,7 @@ public static partial class TypeExtensions
         public static System.Reflection.FieldInfo[] GetFields([System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.NonPublicFields | System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.PublicFields)] this System.Type type, System.Reflection.BindingFlags bindingAttr) { throw null; }
         public static System.Type[] GetGenericArguments(this System.Type type) { throw null; }
-        public static System.Type[] GetInterfaces(this System.Type type) { throw null; }
+        public static System.Type[] GetInterfaces([System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute(System.Diagnostics.CodeAnalysis.DynamicallyAccessedMemberTypes.Interfaces)] this System.Type type) { throw null; }
```
Did we update the `System.Runtime` "ref" assembly?

vitek-karas committed:
PR feedback

vitek-karas committed:
Fix ref assemblies

eerhardt reviewed (approved):
LGTM. I had just one more comment on TypeInfo.ImplementedInterfaces ref file. Other than that, looks good.

vitek-karas committed:
Fix ref assemblies

vitek-karas committed:
Annotate return value of GetInterface with Interfaces.

The Interfaces annotation is transitive (by definition), so once it applies to a Type it also applies to all the interfaces implemented by that type.

vitek-karas committed:
Fix build issues on Linux

eerhardt reviewed (approved):
:shipit:
